felix/fv/bpf_programs_test.go

```diff
@@ -92,4 +94,48 @@ var _ = infrastructure.DatastoreDescribe("_BPF-SAFE_ Felix bpf reattach object",
 		Eventually(secondRunProg2, "10s", "100ms").Should(BeClosed())
 		Expect(secondRunBase).NotTo(BeClosed())
 	})
+
+	It("should not leak bpf programs", func() {
+		noOfBPFPRograms := func() int {
+			out, err := felixes[0].ExecCombinedOutput("bpftool", "prog", "list")
+			Expect(err).NotTo(HaveOccurred())
+			noOfProgs := strings.Count(out, "sched_cls")
+			return noOfProgs
+		}
+
+		const noOfWorkloads = 5
+		var wl [noOfWorkloads]*workload.Workload
+		initVal := noOfBPFPRograms()
+		expectedWorkloadPrograms := 2*5 + 2 // 5 programs for each hook + 2 policy programs
+		expectedHostPrograms := 2 * 5       // 5 programs for each hook
+
+		felixes[0].TriggerDelayedStart()
+		Eventually(noOfBPFPRograms, "10s", "100ms").Should(Equal(initVal + expectedHostPrograms))
+
+		By("creating workloads")
+		for i := 0; i < noOfWorkloads; i++ {
+			wl[i] = workload.New(felixes[0], fmt.Sprintf("w%d", i), "default", fmt.Sprintf("10.65.0.%d", i+2), "8085", "tcp")
+			wl[i].Start()
+			wl[i].ConfigureInInfra(infra)
+			Eventually(noOfBPFPRograms, "10s", "100ms").Should(Equal(initVal + (i+1)*expectedWorkloadPrograms + expectedHostPrograms))
+		}
+
+		By("restarting workloads")
+		totalProgs := noOfBPFPRograms()
+		for i := 0; i < noOfWorkloads; i++ {
+			wl[i].Stop()
+			Eventually(noOfBPFPRograms, "10s", "100ms").Should(Equal(totalProgs - expectedWorkloadPrograms))
+			wl[i].Start()
+			Eventually(noOfBPFPRograms, "10s", "100ms").Should(Equal(totalProgs))
+		}
+
+		By("removing workloads")
+		for i := 0; i < noOfWorkloads; i++ {
+			wl[i].Stop()
+			wl[i].RemoveFromInfra(infra)
+			Eventually(noOfBPFPRograms, "10s", "100ms").Should(Equal(totalProgs - (i+1)*expectedWorkloadPrograms))
```

Comment:
Since we have tested the churn in the above block, I think it is ok to do a single check after all workloads are removed. WDYT?